File: Webclient/src/app/module/viewer/domain_value/spell_component.ts
import {School} from "./school";
import {Mitigation} from "./mitigation";

export interface SpellComponent {
    school_mask: Array<School>;
    amount: number;
    mitigation: Array<Mitigation>;
}

export function get_spell_components_total_amount(components: Array<SpellComponent>): number {
    return components.reduce((acc, comp) => acc + comp.amount, 0);
}

File: Backend/src/modules/live_data_processor/domain_value/spell_component.rs
use crate::modules::live_data_processor::domain_value::{Mitigation, SchoolMask};

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct SpellComponent {
    pub school_mask: SchoolMask,
    pub amount: u32,
    pub mitigation: Vec<Mitigation>,
}

pub fn get_spell_components_total(components: &Vec<SpellComponent>) -> u32 {
    components.iter().fold(0, |acc, comp| {
        acc + comp.amount
            + comp
                .mitigation
                .iter()
                .find_map(|mitigation| {
                    if let Mitigation::Absorb(amount) = mitigation {
                        return Some(*amount);
                    }
                    None
                })
                .unwrap_or(0)
    })
}
